spot_driver/src/api/default_image_client.cpp

```diff
@@ -134,13 +139,21 @@ tl::expected<sensor_msgs::msg::Image, std::string> toImageMsg(const bosdyn::api:
     // First we create a cv::Mat which contains the compressed image data...
     const cv::Mat img_compressed{1, image.rows() * image.cols(), CV_8UC1, &data.front()};
     // Then we decode it to extract the raw image into a new cv::Mat.
-    // Note: this assumes that if an image is provided as JPEG-compressed data, then it is an RGB image.
-    const cv::Mat img_bgr = cv::imdecode(img_compressed, cv::IMREAD_COLOR);
-    if (!img_bgr.data) {
-      return tl::make_unexpected("Failed to decode JPEG-compressed image.");
+    if (image.pixel_format() == bosdyn::api::Image_PixelFormat_PIXEL_FORMAT_GREYSCALE_U8) {
+      const cv::Mat img_grey = cv::imdecode(img_compressed, cv::IMREAD_GRAYSCALE);
+      if (!img_grey.data) {
+        return tl::make_unexpected("Failed to decode JPEG-compressed image.");
+      }
+      const auto image = cv_bridge::CvImage{header, "mono8", img_grey}.toImageMsg();
+      return *image;
+    } else {
+      const cv::Mat img_bgr = cv::imdecode(img_compressed, cv::IMREAD_COLOR);
+      if (!img_bgr.data) {
+        return tl::make_unexpected("Failed to decode JPEG-compressed image.");
+      }
+      const auto image = cv_bridge::CvImage{header, "bgr8", img_bgr}.toImageMsg();
+      return *image;
```

Comment:
So the published data should depend on how the robot is configured via the provided control interface in the SPOT-provided API. ROS driver should allow the same configuration and the 1:1 forwarding of the data to ROS. 
PS: At least that's what I now did for the native ROS2 interface in the AD-Driving Simulator CARLA: try to provide the rendering output from Unreal Engine 1:1 towards ROS if possible and adapt only where necessary (e.g. adapt data format to pointCloud ros-message).

Reply:
Hi @berndgassmann,
Our plan with this PR is to add an optional node that will decompress the images in case the user wants that built-in. This will also minimize the impact on our internal projects with Spot that are assuming to receive uncompressed images. I agree that it's best to have the images always published as compressed by default!
As far as I'm aware, there's no way to get the motor currents from Spot. The closest thing to what you mentioned seems to be the topic `/<robot name>/status/feet`, which gives the position of the feet with respect to the body and whether or not they are in contact with the ground. You could reach out to Boston Dynamics support as well and they might have more insight on this.

Reply:
Ah thanks, yes; that's what I am searching for indeed. That should work out. It even has already some estimates on slipping and friction, so I don't need my fuzzy estimators for such. Maybe I'll then try my odometry plus the Kalman filter on top for orientation correction ImuState (and potentially Gps or other fixed absolute position information to correct the position error if I get the muse ;-)